Bottom: 6fa57f3d012a8daa9ae7e14e971a0635a2aca38b
Top:    0bba04a5b5873e767101bde675e5606dcb788118
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2015-02-05 21:14:33 +0000

Refresh of polygon-debug-operations

---

diff --git a/src/main.c b/src/main.c
index 2e7c365..97bc786 100644
--- a/src/main.c
+++ b/src/main.c
@@ -1875,6 +1875,35 @@ char *program_directory = 0;
 
 #include "dolists.h"
 
+static POLYAREA *
+LayerPolygonsToPoly (LayerType *layer)
+{
+  GList *iter;
+  POLYAREA *piece;
+  POLYAREA *ret = NULL;
+
+  for (iter = layer->Polygon;
+       iter != NULL;
+       iter = g_list_next (iter))
+    {
+      piece = PolygonToPoly (iter->data);
+
+      if (ret == NULL)
+        {
+          ret = piece;
+        }
+      else
+        {
+          piece->f = ret;
+          piece->b = ret->b;
+          ret->b->f = piece;
+          ret->b = piece;
+        }
+    }
+
+  return ret;
+}
+
 int
 main (int argc, char *argv[])
 {
@@ -2010,27 +2039,24 @@ main (int argc, char *argv[])
        * file might not exist
        */
       if (LoadPCB (command_line_pcb))
+	PCB->Filename = strdup (command_line_pcb);
+
+      /* Polygon debug hack - perform an operation if there is a polygon on the first two layers */
+      if (max_copper_layer >= 3 &&
+          LAYER_PTR(0)->PolygonN >= 1 &&
+          LAYER_PTR(1)->PolygonN >= 1 &&
+          LAYER_PTR(2)->PolygonN == 0)
         {
-          PCB->Filename = strdup (command_line_pcb);
-
-          /* Polygon debug hack - perform an operation if there is a polygon on the first two layers */
-          if (max_copper_layer >= 3 &&
-              LAYER_PTR(0)->PolygonN == 1 &&
-              LAYER_PTR(1)->PolygonN == 1 &&
-              LAYER_PTR(2)->PolygonN == 0)
-            {
-              POLYAREA *a, *b, *res;
-              int operation;
-
-              a = PolygonToPoly ((PolygonType *)LAYER_PTR(0)->Polygon->data);
-              b = PolygonToPoly ((PolygonType *)LAYER_PTR(1)->Polygon->data);
-              operation = PBO_UNITE;
-
-              poly_Boolean_free (a, b, &res, operation);
-
-              if (res != NULL)
-                PolyToPolygonsOnLayer (PCB->Data, LAYER_PTR(2), res, NoFlags());
-            }
+          POLYAREA *a, *b, *res;
+          int operation = PBO_UNITE;
+
+          a = LayerPolygonsToPoly (LAYER_PTR(0));
+          b = LayerPolygonsToPoly (LAYER_PTR(1));
+
+          poly_Boolean_free (a, b, &res, operation);
+
+          if (res != NULL)
+            PolyToPolygonsOnLayer (PCB->Data, LAYER_PTR(2), res, NoFlags());
         }
     }
